### ISSUE TITLE
graph: [proposal] add algorithm for finding intervals in flow graphs

### ISSUE THREAD

**Issue by mewmew** [gonum/gonum#796]
### Background

There exist a number of core concepts in graph theory that are general enough to be useful outside of specific application use cases. Concepts such as strongly connected regions, shortest path, dominance relationships in flow graphs, etc.

Related to flow graphs there is the notion of intervals, which are similar to strongly connected regions, but differ in definition and thus application.

The definition of a an interval is as follows:

From Allen, Frances E. "Control flow analysis." ACM Sigplan Notices. Vol. 5. No. 7. ACM, 1970. (ref: http://www.cs.columbia.edu/~suman/secure_sw_devel/p1-allen.pdf)

> Given a node *h*, an interval *I(h)* is the maximal, single entry subgraph for which *h* is the entry node and in which all closed paths contain *h*. The unique interval node *h* is called the interval head or simply the header node.

![interval](https://user-images.githubusercontent.com/1414531/51094144-de1f8e80-17aa-11e9-96d5-fd5ca68370d1.png)

### Proposal

Implement algorithm for finding intervals in rooted flow graphs. The specific algorithm is outlined in these papers (with slight variations between the papers, but the output is identical).

* Allen, Frances E., and John Cocke. "A program data flow analysis procedure." Communications of the ACM 19.3 (1976): 137.
* Figure 1 "Interval Algorithm" in C. Cifuentes, "A Structuring Algorithm for Decompilation", 1993.
* Figure 6-8 "Interval Algorithm" in C. Cifuentes, "Reverse Comilation Techniques", 1994.

### Potential impact of proposal

It should be enough to introduce the following API to the `graph/path` package, which already contains control flow graph algorithms (e.g. calculation of dominance trees).

Of course, these API additions are open for suggestions. Mainly plotting them down, so we can get a feel for what it may look like.

```go
// Intervals returns the unique set of intervals in the given control flow
// graph rooted at the node with ID eid.
func Intervals(g graph.Directed, eid int64) []*Interval {
}

// An Interval I(h) is the maximal, single entry subgraph for which h
// is the entry node and in which all closed paths contain h. The
// unique interval node h is called the interval head or simply the
// header node.
type Interval struct {
   // Interval head.
   head graph.Node
   // Interval nodes.
   nodes []graph.Node
}

// Head returns the header node of the interval.
func (i *Interval) Head() graph.Node {
}

// Nodes returns all the nodes in the interval.
func (i *Interval) Nodes() graph.Nodes {
}

// Edge returns the edge from u to v, with IDs uid and vid,
// if such an edge exists between the nodes of the interval
// and nil otherwise. The node v must be directly reachable
// from u as defined by the From method.
func (i *Interval) Edge(uid, vid int64) graph.Edge {
}
```

The intention is to have `*Interval` implement the `graph.Directed` interface.

Thoughts and comments? I'd be submitting a PR with the implementation, once we've settled on if this is an algorithm we wish to include in the Gonum graph package, and once we've discussed what the API should look like.

Cheers,
Robin

**Comment by mewmew** [comment]
In relation to this, the concept of a rooted flow graph is useful and prevalent. Perhaps it makes sense to make it a first-class citizen in Gonum graph?

This could update e.g. `path.Dominators` as follows:

```diff
-func Dominators(root graph.Node, g graph.Directed) DominatorTree
+func Dominators(g FlowGraph) DominatorTree
```

Where `FlowGraph` is defined as follows:

```go
// FlowGraph is a rooted flow graph.
type FlowGraph interface {
   graph.Directed
   // Root returns the root node of the flow graph.
   Root() graph.Node
}
```

**Comment by kortschak** [comment]
I'll read this more fully later, but something to consider is the possibility of a move to another package. Putting the dominator analysis code into path always felt a little weird to me, but in the absence of more CF code, that's where it ended up. With an extended CF suite, we should consider adding a new package for it; "graph/cfa"? "graph/control"? "graph/flow"? I'm not sure.

**Comment by mewmew** [comment]
Sure, we can create a more dedicated package for control flow analysis. I think both `cfa` and `flow` are good package names.

**Comment by kortschak** [comment]
I think "flow" is OK. What I am concerned about is that there are also dynamic flow graphs (actually [flow graphs](https://en.wikipedia.org/wiki/Flow_graph_(mathematics))). Calling it flow though would have the nice property that your proposed `FlowGraph` could just be `flow.Graph`.

I think either "graph/cfg" or "graph/control/flow" (likely with no other package in "control", making this less appealing) ("graph/conflow"?). The other possibility is that flow graphs of both types (being closely related) end up in the same package.

Apart from all that, this seems eminently reasonable.

**Comment by mewmew** [comment]
> I think either "graph/cfg" or "graph/control/flow" (likely with no other package in "control", making this less appealing) ("graph/conflow"?). The other possibility is that flow graphs of both types (being closely related) end up in the same package.

To get this started, I'll use `cfa` for now. We can easily change this name once we settle on something better.

**Comment by kortschak** [comment]
SGTM

**Comment by T14Raptor** [comment]
Any update on this?

**Comment by mewmew** [comment]
> Any update on this?

Hi @T14Raptor!

No updates from my side. I would love to get back to it in the future, but that will not be in the near future as I'm otherwise engaged with other exciting explorations : )

Wish you all the best though, and if you would like to "take up the torch" and continue the development, you're warmly invited.

With kindness,
Robin

**Comment by harpend** [comment]
Hi @mewmew,

Just working on this same problem for my own project, would be more than happy to tackle it once I've figured it out if you're still interested?

Thanks,
Jack

**Comment by mewmew** [comment]
> Hi @mewmew,
> 
> Just working on this same problem for my own project, would be more than happy to tackle it once I've figured it out if you're still interested?
> 
> Thanks, Jack

Hi Jack!

Happy to hear you're also working on the same kind of problems. What project is it you're working on if I may ask? : )

Finding intervals in flow graphs is very useful for a lot of control flow analysis, and I'd love to see this in gonum.

My personal project for control flow recovery is on a hiatus. That being said, if you get the ball rolling, I'd be glad to join in the discussion to get this integrated into gonum.

Happy regards,
Robin

**Comment by harpend** [comment]
> Hi Jack!
> 
> Happy to hear you're also working on the same kind of problems. What project is it you're working on if I may ask? : )
> 
> Finding intervals in flow graphs is very useful for a lot of control flow analysis, and I'd love to see this in gonum.
> 
> My personal project for control flow recovery is on a hiatus. That being said, if you get the ball rolling, I'd be glad to join in the discussion to get this integrated into gonum.
> 
> Happy regards, Robin

Sounds good, I should note I'm relatively busy at the moment.

The project I'm building is a Java decompiler, I can see your personal project is also one!

Thanks,
Jack

**Comment by mewmew** [comment]
> > Hi Jack!
> > Happy to hear you're also working on the same kind of problems. What project is it you're working on if I may ask? : )
> > Finding intervals in flow graphs is very useful for a lot of control flow analysis, and I'd love to see this in gonum.
> > My personal project for control flow recovery is on a hiatus. That being said, if you get the ball rolling, I'd be glad to join in the discussion to get this integrated into gonum.
> > Happy regards, Robin
> 
> Sounds good, I should note I'm relatively busy at the moment.
> 
> The project I'm building is a Java decompiler, I can see your personal project is also one!
> 
> Thanks, Jack

That's awesome! Really cool to see you also working on decompilers : ) Would be happy to also collaborate in the future and share some common data flow analysis and control flow analysis methods.

Cheerful regards,
Robin

**Comment by harpend** [comment]
Hi @mewmew,

do you think it would be more appropriate to have:
func IntervalGraph(g graph.Directed, eid int64) graph.Directed {
}

over
func Intervals(g graph.Directed, eid int64) []*Interval {
}

as this means it could be used recursively until it is no longer reducible or reaches one node.

Let me know what you think?

Thanks,
Jack

**Comment by harpend** [comment]
Or maybe a separate function
func IntervalsToGraph([]*Interval, g graph.Directed) graph.Directed {
}

**Comment by mewmew** [comment]
Hmm, lets see if I understand (or maybe I missed something).

>  `func IntervalGraph(g graph.Directed, eid int64) graph.Directed {`

There may be one or more intervals in a given control flow graph, that's why the original function signature returned a slice of intervals, instead of a single directed graph.

Given that `*Interval` implements the `graph.Directed` interface, should the original function signature not be enough for recursive calls?

re:
> The intention is to have `*Interval` implement the `graph.Directed` interface.

**Comment by harpend** [comment]
Yes this would be enough, my thoughts were more in line with the decompilation idea of creating an interval graph with links between the intervals and continuing to reduce these graphs to determine the nesting of loops. 

Example (C. Cifuentes, "Reverse Compilation Techniques", 1994) :
![image](https://github.com/user-attachments/assets/b655fcba-a85e-4b9f-a79e-027402a01b2f)

Therefore the idea would be each interval itself is a directed graph and they combine together to make another directed graph.

However, it is more difficult to implement this.

Other than that if the idea is to have the intervals each as separate graphs I should be able to be done today/tommorow.

**Comment by kortschak** [comment]
If necessary, a recomposition type that has at least a `[]*Interval` field should suffice for that.
